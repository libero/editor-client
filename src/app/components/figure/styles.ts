import { makeStyles } from '@material-ui/core/styles';
import { CSSProperties } from '@material-ui/core/styles/withStyles';

export const useFigureEditorStyles = makeStyles((theme) => ({
  figureContainer: {
    display: 'flex',
    alignItems: 'flex-start',

    '& .MuiOutlinedInput-root.Mui-focused fieldset': {
      borderWidth: 1
    }
  },
  figureContent: {
    flexGrow: 1,
    padding: theme.spacing(3, 2, 2, 2),
    margin: theme.spacing(5, 0),
    borderRadius: theme.shape.borderRadius,
    borderColor: 'rgba(0, 0, 0, 0.23)',
    borderStyle: 'solid',
    borderWidth: '1px'
  },
  deleteButton: {
    flexGrow: 0,
    padding: 6,
    marginLeft: theme.spacing(2),
    marginTop: theme.spacing(5)
  },
  inputField: {
    marginBottom: theme.spacing(2.5)
  },
  imageContainer: {
    marginBottom: theme.spacing(2.5),
    position: 'relative'
  },
  uploadImageCta: {
    position: 'absolute',
    top: theme.spacing(1),
    right: theme.spacing(1),
    backgroundColor: '#929393',
    padding: 10,
    '&:hover': {
      backgroundColor: '#929393'
    },
    '& svg': {
      fill: '#FFFFFF'
    }
  },
  image: {
    maxWidth: '100%',
    minHeight: 58,
    borderRadius: theme.shape.borderRadius,
    border: 'solid 1px #C4C4C4'
  }
}));

export const useFigureLicenseEditorStyles = (theme): Record<string, CSSProperties> => ({
  editorContainer: {
    display: 'flex'
  },
  editorForm: {
    borderRadius: theme.shape.borderRadius,
    borderColor: 'rgba(0, 0, 0, 0.23)',
    borderStyle: 'solid',
    borderWidth: '1px',
    marginRight: theme.spacing(1),
    padding: theme.spacing(1.5)
  },
  smallField: {
    width: 150,
    marginRight: theme.spacing(2)
  },
  deleteButton: {
    flexGrow: 0,
    padding: 6,
    alignSelf: 'flex-start'
  },
  spacer: {
    flexGrow: 1
  },
  inputField: {
    marginBottom: theme.spacing(2.5),
    '&:last-child': {
      marginBottom: 0
    }
  },
  fieldsRow: {
    display: 'flex',
    marginBottom: theme.spacing(2.5)
  }
});
