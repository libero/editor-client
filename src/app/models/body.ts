import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
