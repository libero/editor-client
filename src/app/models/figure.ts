import { get } from 'lodash';

import { getTextContentFromPath } from 'app/models/utils';

export interface Figure {
  copyrightHolder: string;
  copyrightStatement: string;
  copyrightYear: string;
  licenseType: string;
}

export const FIGURE_LICENSE_CC0 = 'LICENSE_CCO';
export const FIGURE_LICENSE_CC_BY_4 = 'LICENSE_CC_BY_4';
export const FIGURE_LICENSE_OTHER = 'LICENSE_OTHER';

export const FIGURE_LICENSE_SELECT_OPTIONS = [
  { label: 'CC0', value: FIGURE_LICENSE_CC0 },
  { label: 'CC by 4.0', value: FIGURE_LICENSE_CC_BY_4 },
  { label: 'Other', value: FIGURE_LICENSE_CC_BY_4 }
];

const LICENSE_URLS_MAP = {
  'http://creativecommons.org/licenses/by/4.0/': FIGURE_LICENSE_CC_BY_4,
  'http://creativecommons.org/publicdomain/zero/1.0/': FIGURE_LICENSE_CC0
};

export function createFigureLicenseAttributes(el: Element): Figure {
  return {
    copyrightHolder: getTextContentFromPath(el, 'copyright-holder'),
    copyrightStatement: getTextContentFromPath(el, 'copyright-statement'),
    copyrightYear: getTextContentFromPath(el, 'copyright-year'),
    licenseType: getLicenseType(el)
  };
}

export function createEmptyLicenseAttributes(): Figure {
  return {
    copyrightHolder: '',
    copyrightStatement: '',
    copyrightYear: '',
    licenseType: undefined
  };
}

export function getFigureImageUrl(el: Element): string {
  const paths = get(el.ownerDocument, 'manuscriptPath').split('/');
  const id = paths[2];

  return process.env.REACT_APP_NO_SERVER
    ? get(el.ownerDocument, 'manuscriptPath') + '/' + get(el.querySelector('graphic'), 'attributes.xlink:href.value')
    : `/api/v1/articles/${id}/assets/${get(el.querySelector('graphic'), 'attributes.xlink:href.value').replace(
        'tif',
        'jpg'
      )}`;
}

function getLicenseType(el: Element): string {
  const licenseUrl =
    getTextContentFromPath(el, 'license_ref') || el.querySelector('license').getAttribute('xlink:href') || '';

  return LICENSE_URLS_MAP[licenseUrl] || FIGURE_LICENSE_OTHER;
}
