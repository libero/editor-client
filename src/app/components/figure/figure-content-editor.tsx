import React from 'react';
import { withStyles, WithStyles } from '@material-ui/core/styles';

import { NodeEditor } from 'app/components/node-editor/node-editor';
import { RichTextEditor } from 'app/components/rich-text-editor';
import { useFigureContentEditorStyles } from 'app/components/figure/styles';

interface FigureContentEditorProps extends WithStyles<typeof useFigureContentEditorStyles> {
  label: string;
}

export class FigureContentEditorComponent extends NodeEditor<FigureContentEditorProps> {
  render() {
    return (
      <div>
        <RichTextEditor
          ref={this.editorRef}
          classes={{ root: this.props.classes.richTextEditorField }}
          isActive={this.getEditorActiveState()}
          variant="outlined"
          label={this.props.label}
          editorState={this.state.editorState}
          onChange={this.handleInternalEditorStateChange}
          onFocus={this.handleEditorFocus}
          onBlur={this.handleEditorBlur}
        ></RichTextEditor>
      </div>
    );
  }
}

const stylesDecorator = withStyles(useFigureContentEditorStyles, { withTheme: true });
export const FigureContentEditor = stylesDecorator(FigureContentEditorComponent);
