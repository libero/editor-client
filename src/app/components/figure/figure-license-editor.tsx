import React, { ReactNode } from 'react';
import { TextField, IconButton } from '@material-ui/core';
import { withStyles, WithStyles } from '@material-ui/core/styles';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import DeleteIcon from '@material-ui/icons/Delete';

import { FIGURE_LICENSE_CC0, FIGURE_LICENSE_SELECT_OPTIONS, Figure } from '../../models/figure';
import { RichTextEditor } from '../rich-text-editor';
import { NodeEditor, NodeEditorState } from '../node-editor/node-editor';
import { useFigureLicenseEditorStyles } from './styles';
import { Select } from '../select';
import { NodeSelection } from 'prosemirror-state';

interface FigureLicenseEditorProps extends WithStyles<typeof useFigureLicenseEditorStyles> {
  node: ProsemirrorNode;
  offset: number;
  index: number;
  classes: Record<string, string>;
}

class FigureContentEditorComponent extends NodeEditor<FigureLicenseEditorProps> {
  shouldComponentUpdate(nextProps: FigureLicenseEditorProps, nextState: NodeEditorState): boolean {
    if (this.props.index !== nextProps.index) {
      this.forceUpdate();
      return true;
    }

    return super.shouldComponentUpdate(nextProps, nextState);
  }

  render(): ReactNode {
    const license: Figure = this.props.node.attrs.licenseInfo as Figure;
    const classes = this.props.classes;
    return (
      <section className={classes.editorContainer}>
        <div className={classes.editorForm}>
          <div className={classes.fieldsRow}>
            <Select
              className={classes.smallField}
              name="licenseType"
              placeholder="Please select"
              canUnselect={false}
              blankValue={undefined}
              label="Licence type"
              options={FIGURE_LICENSE_SELECT_OPTIONS}
              value={license.licenseType}
              onChange={this.handleFormChange}
            />
            {license.licenseType !== FIGURE_LICENSE_CC0 ? (
              <TextField
                classes={{ root: classes.smallField }}
                name="copyrightYear"
                label="Year"
                InputLabelProps={{ shrink: true }}
                variant="outlined"
                value={license.copyrightYear}
                onChange={this.handleFormChange}
              />
            ) : undefined}
          </div>
          {license.licenseType !== FIGURE_LICENSE_CC0 ? (
            <TextField
              fullWidth
              name="copyrightStatement"
              label={`License ${this.props.index + 1} statement`}
              classes={{ root: classes.inputField }}
              InputLabelProps={{ shrink: true }}
              variant="outlined"
              value={license.copyrightStatement}
              onChange={this.handleFormChange}
            />
          ) : undefined}
          <div className={classes.inputField}>
            <RichTextEditor
              ref={this.editorRef}
              classes={{ root: this.props.classes.richTextEditorField }}
              isActive={this.state.isEditorActive}
              label={`License ${this.props.index + 1}`}
              variant="outlined"
              editorState={this.state.editorState}
              onChange={this.handleInternalEditorStateChange}
              onFocus={this.handleEditorFocus}
              onBlur={this.handleEditorBlur}
            />
          </div>
          {license.licenseType !== FIGURE_LICENSE_CC0 ? (
            <TextField
              fullWidth
              name="copyrightHolder"
              label={`License ${this.props.index + 1} holder`}
              InputLabelProps={{ shrink: true }}
              variant="outlined"
              value={license.copyrightHolder}
              onChange={this.handleFormChange}
            />
          ) : undefined}
        </div>
        <IconButton classes={{ root: classes.deleteButton }} onClick={this.handleDelete}>
          <DeleteIcon fontSize="small" />
        </IconButton>
      </section>
    );
  }

  private handleFormChange = (event): void => {
    const newAttributes = {
      id: this.props.node.attrs.id,
      licenseInfo: {
        ...this.props.node.attrs.licenseInfo,
        [event.target['name']]: event.target['value']
      }
    };
    const change = this.context.view.state.tr;
    change.setNodeMarkup(this.context.getPos() + this.props.offset - 1, null, newAttributes);
    this.context.view.dispatch(change);
  };

  private handleDelete = (): void => {
    const licenseNodePosition = this.context.view.state.doc.resolve(this.context.getPos() + this.props.offset - 1);
    const change = this.context.view.state.tr.setSelection(new NodeSelection(licenseNodePosition)).deleteSelection();
    this.context.view.dispatch(change);
  };
}

const stylesDecorator = withStyles(useFigureLicenseEditorStyles, { withTheme: true });
export const FigureLicenseEditor = stylesDecorator(FigureContentEditorComponent);
