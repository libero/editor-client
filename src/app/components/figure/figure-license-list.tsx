import React, { useCallback, useContext } from 'react';
import { Button } from '@material-ui/core';
import { Node as ProsemirrorNode } from 'prosemirror-model';

import { FigureLicenseEditor } from 'app/components/figure/figure-license-editor';
import { NodeViewContext } from 'app/utils/view.utils';
import { createEmptyLicenseAttributes } from 'app/models/figure-license';

interface FigureLicenseListProps {
  licenses: Array<{
    node: ProsemirrorNode;
    offset: number;
  }>;
}

const FIGURE_LICENSE_NODE_OFFSET_CORRECTION = 2;

export const FigureLicensesList: React.FC<FigureLicenseListProps> = (props) => {
  const { licenses } = props;
  const nodeViewContext = useContext(NodeViewContext);

  const handleAddNewLicense = useCallback(() => {
    const newNodePosition = nodeViewContext.getPos() + nodeViewContext.node.nodeSize - 1;
    const newNode = nodeViewContext.view.state.schema.nodes.figureLicense.createAndFill({
      licenseInfo: createEmptyLicenseAttributes()
    });
    const change = nodeViewContext.view.state.tr.insert(newNodePosition, newNode);
    nodeViewContext.view.dispatch(change);
  }, [nodeViewContext]);

  return (
    <section>
      {licenses.map(({ node, offset }, index) => {
        return (
          <FigureLicenseEditor
            key={index}
            node={node}
            offset={offset + FIGURE_LICENSE_NODE_OFFSET_CORRECTION}
            index={index}
          />
        );
      })}
      <Button onClick={handleAddNewLicense} color="primary">
        Add Licence
      </Button>
    </section>
  );
};
