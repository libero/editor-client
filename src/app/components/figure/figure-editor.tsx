import React, { useCallback, useRef, useImperativeHandle, useState } from 'react';
import { EditorState, Transaction, Selection } from 'prosemirror-state';
import { Node as ProsemirrorNode, NodeSpec } from 'prosemirror-model';
import DeleteIcon from '@material-ui/icons/Delete';
import { IconButton, TextField } from '@material-ui/core';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { get } from 'lodash';

import { RichTextEditor } from 'app/components/rich-text-editor';
import { useFigureEditorStyles } from 'app/components/figure/styles';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { SelectPlugin } from 'app/models/plugins/selection.plugin';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';

export interface FigureEditorHandle {
  updateContent(node: ProsemirrorNode): void;
  focusFromSelection(selection: Selection, figurePos: number): void;
  hasFocus(): boolean;
}

interface FigureEditorProps {
  figureNode: ProsemirrorNode;
  onDelete(): void;
  onNodeChange(change: Transaction, offset: number): void;
  onLabelChange(label: string): void;
  onSelectionChange(from: number, anchor: number): void;
}

export const FigureEditor = React.forwardRef((props: FigureEditorProps, ref) => {
  const { figureNode, onNodeChange, onSelectionChange, onDelete, onLabelChange } = props;
  const [isTitleEditorActive, setTitleEditorActive] = useState<boolean>(false);
  const [isLegendEditorActive, setLegendEditorActive] = useState<boolean>(false);
  const [label, setLabel] = useState<string>(figureNode.attrs.label);
  const [internalTitleState, setInternalTitleState] = useState<EditorState>(createFigureTitleState(figureNode));
  const [internalLegendState, setInternalLegendState] = useState<EditorState>(createFigureLegendState(figureNode));
  const [titleOffset, setTitleOffset] = useState<number>(
    findChildrenByType(figureNode, figureNode.type.schema.nodes.figureTitle)[0].offset
  );
  const [legendOffset, setLegendOffset] = useState<number>(
    findChildrenByType(figureNode, figureNode.type.schema.nodes.figureLegend)[0].offset
  );
  const classes = useFigureEditorStyles();
  const titleEditorRef = useRef(null);
  const legendEditorRef = useRef(null);

  useImperativeHandle(ref, () => ({
    updateContent: (updatedFigureNode: ProsemirrorNode) => {
      setLabel(updatedFigureNode.attrs.label);
      const updatedTitleNode = findChildrenByType(
        updatedFigureNode,
        updatedFigureNode.type.schema.nodes.figureTitle
      )[0];

      const updatedLegendNode = findChildrenByType(
        updatedFigureNode,
        updatedFigureNode.type.schema.nodes.figureLegend
      )[0];

      const titleChange = getUpdatesForNode(updatedTitleNode.node, titleEditorRef.current.editorView.state);
      if (titleChange) {
        titleEditorRef.current.editorView.dispatch(titleChange);
      }

      const legendChange = getUpdatesForNode(updatedLegendNode.node, legendEditorRef.current.editorView.state);
      if (legendChange) {
        legendEditorRef.current.editorView.dispatch(legendChange);
      }

      setTitleOffset(updatedTitleNode.offset);
      setLegendOffset(updatedLegendNode.offset);
    },
    focusFromSelection: (selection: Selection, figurePos: number) => {
      const cursorPos = selection.$from.pos;
      if (
        figurePos + titleOffset <= cursorPos &&
        cursorPos <= figurePos + titleOffset + internalTitleState.doc.nodeSize
      ) {
        titleEditorRef.current.focus();
      } else if (
        figurePos + legendOffset <= cursorPos &&
        cursorPos <= figurePos + legendOffset + internalLegendState.doc.nodeSize
      ) {
        legendEditorRef.current.focus();
      }
    },
    hasFocus: () => titleEditorRef.current.editorView.hasFocus() || legendEditorRef.current.editorView.hasFocus()
  }));

  const handleLabelChange = useCallback(
    (event) => {
      onLabelChange(event.target.value);
      setLabel(event.target.value);
    },
    [onLabelChange]
  );

  const handleTitleChange = useCallback(
    (change: Transaction) => {
      setInternalTitleState(titleEditorRef.current.editorView.state);
      if (change.docChanged && !change.getMeta('parentChange')) {
        onNodeChange(change, titleOffset);
      } else {
        onSelectionChange(change.selection.$from.pos + titleOffset, change.selection.$to.pos + titleOffset);
      }
    },
    [onNodeChange, onSelectionChange, titleOffset]
  );

  const handleLegendChange = useCallback(
    (change: Transaction) => {
      setInternalLegendState(legendEditorRef.current.editorView.state);
      if (change.docChanged && !change.getMeta('parentChange')) {
        onNodeChange(change, legendOffset);
      } else {
        onSelectionChange(change.selection.$from.pos + legendOffset, change.selection.$to.pos + legendOffset);
      }
    },
    [legendOffset, onNodeChange, onSelectionChange]
  );

  const handleTitleFocus = useCallback(() => {
    setTitleEditorActive(true);
  }, [setTitleEditorActive]);

  const handleTitleBlur = useCallback(() => {
    setTitleEditorActive(false);
  }, [setTitleEditorActive]);

  const handleLegendFocus = useCallback(() => {
    setLegendEditorActive(true);
  }, []);

  const handleLegendBlur = useCallback(() => {
    setLegendEditorActive(false);
  }, []);

  return (
    <div className={classes.figureContainer}>
      <div className={classes.figureContent}>
        <TextField
          fullWidth
          name="figureNumber"
          label="Figure number"
          classes={{ root: classes.inputField }}
          InputLabelProps={{ shrink: true }}
          variant="outlined"
          multiline
          value={label}
          onChange={handleLabelChange}
        />
        <div className={classes.inputField}>
          <img className={classes.image} src={figureNode.attrs.img} />
        </div>
        <div className={classes.inputField}>
          <RichTextEditor
            ref={titleEditorRef}
            isActive={isTitleEditorActive}
            variant="outlined"
            label="Title"
            editorState={internalTitleState}
            onChange={handleTitleChange}
            onFocus={handleTitleFocus}
            onBlur={handleTitleBlur}
          ></RichTextEditor>
        </div>
        <RichTextEditor
          ref={legendEditorRef}
          isActive={isLegendEditorActive}
          variant="outlined"
          label="Legend"
          editorState={internalLegendState}
          onChange={handleLegendChange}
          onFocus={handleLegendFocus}
          onBlur={handleLegendBlur}
        ></RichTextEditor>
      </div>
      <IconButton classes={{ root: classes.deleteButton }} onClick={onDelete}>
        <DeleteIcon fontSize="small" />
      </IconButton>
    </div>
  );
});

function createFigureTitleState(node: ProsemirrorNode): EditorState {
  const titleNode = findChildrenByType(node, node.type.schema.nodes.figureTitle)[0];
  return EditorState.create({
    doc: titleNode.node,
    plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), SelectPlugin, PlaceholderPlugin('')]
  });
}

function createFigureLegendState(node: ProsemirrorNode): EditorState {
  const legendNode = findChildrenByType(node, node.type.schema.nodes.figureLegend)[0];
  return EditorState.create({
    doc: legendNode.node,
    plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), SelectPlugin, PlaceholderPlugin('')]
  });
}

function getUpdatesForNode(updatedNode: ProsemirrorNode, state: EditorState): Transaction | null {
  const start = updatedNode.content.findDiffStart(state.doc.content);
  if (start !== null) {
    let { a: endA, b: endB } = updatedNode.content.findDiffEnd(get(state, 'doc.content'));
    const overlap = start - Math.min(endA, endB);
    if (overlap > 0) {
      endA += overlap;
      endB += overlap;
    }
    return state.tr.replace(start, endB, updatedNode.slice(start, endA)).setMeta('parentChange', true);
  }

  return null;
}

function findChildrenByType(
  node: ProsemirrorNode,
  nodeType: NodeSpec
): Array<{ node: ProsemirrorNode; offset: number }> {
  const foundChildren = [];
  node.descendants((childNode, pos) => {
    if (nodeType === childNode.type) {
      foundChildren.push({ node: childNode, offset: pos });
    }
  });
  return foundChildren;
}
