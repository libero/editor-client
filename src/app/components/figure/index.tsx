import React from 'react';
import { NodeView, EditorView } from 'prosemirror-view';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import ReactDOM from 'react-dom';
import { ThemeProvider } from '@material-ui/core/styles';
import { Transaction, TextSelection, NodeSelection } from 'prosemirror-state';

import { StepMap } from 'prosemirror-transform';

import { theme } from 'app/styles/theme';
import { FigureEditor, FigureEditorHandle } from 'app/components/figure/figure-editor';

export class FigureNodeView implements NodeView {
  dom?: HTMLElement;
  figureEditor: React.RefObject<FigureEditorHandle>;

  constructor(
    private node: ProsemirrorNode,
    private view: EditorView,
    private getPos: () => number,
    private isContainerActive: () => boolean
  ) {
    this.dom = document.createElement('section');
    this.figureEditor = React.createRef();
    ReactDOM.render(
      <ThemeProvider theme={theme}>
        <FigureEditor
          ref={this.figureEditor}
          figureNode={this.node}
          onDelete={this.handleDelete}
          onLabelChange={this.handleLabelChange}
          onImageChange={this.handleImageChange}
          onNodeChange={this.handleContentChange}
          onSelectionChange={this.handleSelectionChange}
        />
      </ThemeProvider>,
      this.dom
    );
  }

  handleLabelChange = (label: string) => {
    const newAttributes = { label, img: this.node.attrs.img };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleImageChange = (img: string) => {
    const newAttributes = { img, label: this.node.attrs.label };
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, newAttributes);
    this.view.dispatch(change);
  };

  handleContentChange = (nodeViewChange: Transaction, offset: number) => {
    this.view.dispatch(this.transformNodeViewChanges(nodeViewChange, offset));
  };

  handleDelete = () => {
    const change = this.view.state.tr
      .setSelection(new NodeSelection(this.view.state.doc.resolve(this.getPos())))
      .deleteSelection();
    this.view.dispatch(change);
  };

  handleSelectionChange = (anchor: number, head: number) => {
    const offset = this.getPos() + 1;
    const start = Math.min(anchor, head);
    const end = Math.max(anchor, head);
    const selection = TextSelection.create(this.view.state.doc, start + offset, end + offset);
    if (!selection.eq(this.view.state.selection)) {
      const selectionChange = this.view.state.tr.setSelection(selection);
      this.view.dispatch(selectionChange);
      this.view.dom.dispatchEvent(new FocusEvent('focus'));
    }
  };

  update(node: ProsemirrorNode) {
    if (!node.sameMarkup(this.node)) {
      this.node = node;
    }
    this.figureEditor.current.updateContent(node);
    if (this.isContainerActive() && !this.figureEditor.current.hasFocus()) {
      this.figureEditor.current.focusFromSelection(this.view.state.selection, this.getPos());
    }
    return true;
  }

  stopEvent(evt) {
    return this.dom.contains(evt.target);
  }

  ignoreMutation() {
    return true;
  }

  private transformNodeViewChanges(change: Transaction, additionalOffset: number): Transaction {
    const parentChange = this.view.state.tr;
    const offsetMap = StepMap.offset(this.getPos() + 2 + additionalOffset);
    const steps = change.steps;
    for (let i = 0; i < steps.length; i++) {
      parentChange.step(steps[i].map(offsetMap));
    }
    return parentChange;
  }
}
