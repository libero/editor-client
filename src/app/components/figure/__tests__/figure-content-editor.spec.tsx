import React from 'react';
import { create } from 'react-test-renderer';

import { FigureContentEditor } from 'app/components/figure/figure-content-editor';
import { createBodyState } from 'app/models/body';

jest.mock('app/components/rich-text-editor', () => ({
  RichTextEditor: () => <div data-cmp="rich-text-editor"></div>
}));

describe('FigureContentEditor', () => {
  it('should render a node', () => {
    const bodyEditorState = createBodyState(document.createElement('div'), '');
    const node = bodyEditorState.schema.nodes.figureLegend.create();

    const wrapper = create(<FigureContentEditor label="test" node={node} offset={0} />);
    expect(wrapper).toMatchSnapshot();
  });
});
