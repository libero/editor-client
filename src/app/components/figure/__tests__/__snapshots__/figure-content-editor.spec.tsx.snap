// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FigureContentEditor should render a node 1`] = `
<div>
  <div
    data-cmp="rich-text-editor"
  />
</div>
`;
